platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -264,6 +264,87 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     }
 }
 
+#pragma mark Offline merge methods
+
+- (void)addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    NSURL *fileURL = [NSURL fileURLWithPath:filePath];
+    
+    [self addContentsOfURL:fileURL withCompletionHandler:completion];
+
+}
+
+- (void)addContentsOfURL:(NSURL *)fileURL withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    
+    if (!fileURL.isFileURL) {
+        [NSException raise:NSInvalidArgumentException format:@"%@ must be a valid file path", fileURL.absoluteString];
+    }
+    if (![fileManager isWritableFileAtPath:fileURL.path]) {
+        [NSException raise:NSInvalidArgumentException format:@"The file path: %@ must be writable", fileURL.absoluteString];
+    }
+    
+    __weak MGLOfflineStorage *weakSelf = self;
+    [self _addContentsOfFile:fileURL.path withCompletionHandler:^(NSArray<MGLOfflinePack *> * _Nullable packs, NSError * _Nullable error) {
+        if (packs) {
+            NSMutableDictionary *packsByIdentifier = [NSMutableDictionary dictionary];
+            
+            MGLOfflineStorage *strongSelf = weakSelf;
+            for (MGLOfflinePack *pack in packs) {
+                [packsByIdentifier setObject:pack forKey:@(pack.mbglOfflineRegion->getID())];
+            }
+            
+            id mutablePacks = [strongSelf mutableArrayValueForKey:@"packs"];
+            NSMutableIndexSet *replaceIndexSet = [NSMutableIndexSet indexSet];
+            [strongSelf.packs enumerateObjectsUsingBlock:^(MGLOfflinePack * _Nonnull pack, NSUInteger idx, BOOL * _Nonnull stop) {
+                MGLOfflinePack *newPack = packsByIdentifier[@(pack.mbglOfflineRegion->getID())];
+                if (newPack) {
+                    MGLOfflinePack *previousPack = [mutablePacks objectAtIndex:idx];
+                    [previousPack invalidate];
+                    [replaceIndexSet addIndex:idx];
+                    [packsByIdentifier removeObjectForKey:@(newPack.mbglOfflineRegion->getID())];
+                }
+
+            }];
+            
+            if (replaceIndexSet.count > 0) {
+                [mutablePacks replaceObjectsAtIndexes:replaceIndexSet withObjects:packs];
+            }
+            
+            [mutablePacks addObjectsFromArray:packsByIdentifier.allValues];
```

Comment:
This line is intended to add any new offline packs to `MGLOfflineStorage.packs`. In that case, `packs` will be longer than `replaceIndexSet` in the call to `-replaceObjectsAtIndexes:withObjects:`, which will probably trigger an exception. (We should assert that the two have the same length beforehand.) Even if it doesn’t, this line would add those same packs to the array redundantly.

Instead of this line, is it possible to add more indices to the end of `replaceIndexSet` to account for those extra packs? If not, then we’d need to build up a separate array of packs that are getting replaced in the `-enumerateObjectsUsingBlock:` call above, then group the calls to `-replaceObjectsAtIndexes:withObjects:` and `-addObjectsFromArray:` to avoid unnecessary KVO notifications.

This is a good scenario to test, by the way. XCTest has conveniences for testing KVO notifications.